pkg/auth/clientinterceptors.go

```diff
@@ -19,50 +19,36 @@ package auth
 import (
 	"context"
 
-	"google.golang.org/grpc/metadata"
-
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
 )
 
 // ClientStreamInterceptor gRPC client interceptor for streams
 func ClientStreamInterceptor(token string) func(context.Context, *grpc.StreamDesc, *grpc.ClientConn, string, grpc.Streamer, ...grpc.CallOption) (grpc.ClientStream, error) {
 	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
-		opts = append(opts, grpc.PerRPCCredentials(TokenAuthStruct{
-			Token: token,
-		}))
+		ctx = updateAuthHeader(ctx, token)
 		return streamer(ctx, desc, cc, method, opts...)
 	}
 }
 
 // ClientUnaryInterceptor gRPC client interceptor for unary methods
 func ClientUnaryInterceptor(token string) func(context.Context, string, interface{}, interface{}, *grpc.ClientConn, grpc.UnaryInvoker, ...grpc.CallOption) error {
 	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
-		opts = append(opts, grpc.PerRPCCredentials(TokenAuthStruct{
-			Token: token,
-		}))
+		ctx = updateAuthHeader(ctx, token)
 		return invoker(ctx, method, req, reply, cc, opts...)
 	}
 }
 
-// TokenAuthStruct authentication token data structure
-type TokenAuthStruct struct {
-	Token string
-}
-
-// GetRequestMetadata callback which returns the Bearer token to be set in request metadata
-func (t TokenAuthStruct) GetRequestMetadata(ctx context.Context, in ...string) (map[string]string, error) {
-	var token string
+func updateAuthHeader(ctx context.Context, token string) context.Context {
 	if md, ok := metadata.FromOutgoingContext(ctx); ok && len(md.Get("authorization")) > 0 {
+		// Token from existing metadata overwrites the one stored in
```

Comment:
existent token shouldn't be replaced by the one provided as parameter then? if so, I'd suggest changing the comment from `stored in` by `provided in current call` or something like that. And the method could do a fast return if no change is made into the metadata.

However, I would expect the provided as parameter having more precedence actually...

Reply:
Clarified comments a bit.

This is a very fragile peace of code. Most of the work was to find a way to update it without breaking tests and compatibility with existing imuclient. We should definitely clean up the code setting up interceptors when thinking about the new version of client libraries.